src/System/Hworker.hs

```diff
@@ -139,6 +148,45 @@ data JobData t = JobData UTCTime t
 -- intermittent problems).
 data ExceptionBehavior = RetryOnException | FailOnException
 
+type JobID = Text
+
+-- | A unique identifier for grouping jobs together.
+newtype BatchID = BatchID UUID deriving (ToJSON, FromJSON, Eq, Show)
+
+data BatchStatus = BatchQueuing | BatchProcessing | BatchFinished
+  deriving (Eq, Show)
+
+encodeBatchStatus :: BatchStatus -> ByteString
+encodeBatchStatus BatchQueuing    = "queueing"
+encodeBatchStatus BatchProcessing = "processing"
+encodeBatchStatus BatchFinished   = "finished"
+
+decodeBatchStatus :: ByteString -> Maybe BatchStatus
+decodeBatchStatus "queueing"   = Just BatchQueuing
+decodeBatchStatus "processing" = Just BatchProcessing
+decodeBatchStatus "finished"   = Just BatchFinished
+decodeBatchStatus _            = Nothing
+
+data BatchJob =
+  BatchJob
+    { batchID        :: BatchID
+    , batchTotal     :: Int
+    , batchCompleted :: Int
+    , batchSuccesses :: Int
+    , batchFailures  :: Int
+    , batchRetries   :: Int
+    , batchStatus    :: BatchStatus
+    }
+
+data JobRef = JobRef JobID (Maybe BatchID) deriving (Eq, Show)
+
+instance ToJSON JobRef where
+  toJSON (JobRef j b) = A.object ["j" .= j, "b" .= b]
+
+instance FromJSON JobRef where
+  parseJSON (A.String j) = pure (JobRef j Nothing)
```

Comment:
This line can eventually go away, presumably, as it exists for migration (I missed it at first, and thought that when you released this, things would be broken!). Maybe a comment to explain _why_ it exists? As otherwise, slightly confusing!

Reply:
Yup, that can definitely go away eventually. I was just a little paranoid about what would happen if we upgraded and still had jobs with the old encoding still in the Redis queue.